README.md
# cloudbeds-api


Cloudbeds.com API Wrapper for PHP with Laravel Integration

## Installation

```
composer require r4kib/cloudbeds-api
```


## Usage

### Initializing The Class

```php
$cloudbeds = new \R4kib\Cloudbeds\Cloudbeds([
    'clientId'                => 'yourId',          // The client ID assigned to you by Amazon
    'clientSecret'            => 'yourSecret',      // The client password assigned to you by Amazon
    'redirectUri'             => 'yourRedirectUri',  // The return URL you specified for your app on Amazon
    'version'                 => 'v1.1'  // API Version, default v1.1
]);
```
### Initializing The Class (Laravel)

1. Publish the configuration
```
php artisan vendor:publish --config
```
2. Setup API details in the `.env` file
```
CLOUDBEDS_API_CLIENT_ID=yourId
CLOUDBEDS_API_CLIENT_SECRET=yourSecret
CLOUDBEDS_API_REDIRECT_URI=yourRedirectUri
CLOUDBEDS_API_VERSION=v1.1
```
3. get the singletone
```php
$cloudbeds = resolve("R4kib\\Cloudbeds\\Cloudbeds");
```
### OAuth Helper
OAuth portion of this implements [thephpleague/oauth2-client](https://github.com/thephpleague/oauth2-client). So for details you should look over there.

```php
$oauthHelper= $cloudbeds->getOauthHelper();
// Get authorization code
if (!isset($_GET['code'])) {
    
    // Get authorization URL
    $authorizationUrl = $oauthHelper->getAuthorizationUrl();

    // Redirect user to authorization URL
    header('Location: ' . $authorizationUrl);
    exit;
} else {
    // Get access token
        $accessToken = $oauthHelper->getAccessToken(
            'authorization_code',
            [
                'code' => $_GET['code']
            ]
        );

    // Get resource owner
        $resourceOwner = $oauthHelper->getResourceOwner($accessToken);
        
    // Now you can store the results to session etc.
    $_SESSION['accessToken'] = $accessToken;
    $_SESSION['resourceOwner'] = $resourceOwner;

    var_dump(
        $resourceOwner->getID(),
        $resourceOwner->getFirstName(),
        $resourceOwner->getLastName(),
        $resourceOwner->getEmail()
    );
}
```
### Make API request
```php
// $params is [key=>value] array. See cloudbeds.com API documentation to view params.
 $cloudbeds->get('/path',$accessToken,$params);
 $cloudbeds->post('/path',$accessToken,$params);
 $cloudbeds->put('/path',$accessToken,$params);
 $cloudbeds->delete('/path',$accessToken,$params);
```


## License

The MIT License (MIT). Please see [License File](http://github.com/r4kib/cloudbeds-api/blob/master/LICENSE) for more information.


 

CHANGELOG.md
# Cloudbeds.com API Wrapper for PHP with Laravel Integration


test/src/CloudbedsTest.php
<?php
/**
 * Created by PhpStorm.
 * User: rakib
 * Date: 06-May-19
 * Time: 9:11 PM
 */

namespace R4kib\Cloudbeds\Test;

use GuzzleHttp\Client;
use GuzzleHttp\ClientInterface;
use PHPUnit\Framework\TestCase;
use R4kib\Cloudbeds\Cloudbeds;
use R4kib\Cloudbeds\Provider\CloudbedsProvider;
use R4kib\Cloudbeds\Exceptions\CloudbedsAPIException;
use Mockery as m;

class CloudbedsTest extends TestCase
{
    protected  $cloudbeds;
    protected function setUp()
    {
        $this->cloudbeds = new Cloudbeds([
            'clientId' => 'mock_client_id',
            'clientSecret' => 'mock_secret',
            'redirectUri' => 'mock_redirect_uri',
            'version' => 'v1.1']);
    }

    public function testGetOauthHelper()
    {
        $oauthHelper=$this->cloudbeds->getOauthHelper();
        $this->assertInstanceOf(CloudbedsProvider::class,$oauthHelper);
        $this->assertNotNull($oauthHelper);

    }

    public function testGetOptions()
    {
        $options=$this->cloudbeds->getOptions();
        $this->assertArrayHasKey('apiBaseUrl',$options);
        $uri = parse_url($options['apiBaseUrl']);
        $this->assertEquals('/api/v1.1',$uri['path']);
    }

    public function testGetHttpClient()
    {
        $httpClient=$this->cloudbeds->getHttpClient();
        $this->assertInstanceOf(Client::class,$httpClient);
    }

    public function testGet()
    {
        $this->assertTrue(method_exists($this->cloudbeds,'get'));

        $client=$this->mockHttpClient(200);
        $this->cloudbeds->setHttpClient($client);

        $got=$this->cloudbeds->get('mock_uri','mock_access_token',['mock_param'=>'mock_value']);
        $this->assertIsArray($got);
        $this->assertArrayHasKey('success',$got);
    }
    public function testPost()
    {
        $this->assertTrue(method_exists($this->cloudbeds,'post'));

        $client=$this->mockHttpClient(200);
        $this->cloudbeds->setHttpClient($client);

        $got=$this->cloudbeds->post('mock_uri','mock_access_token',['mock_param'=>'mock_value']);
        $this->assertIsArray($got);
        $this->assertArrayHasKey('success',$got);
    }
    public function testDelete()
    {
        $this->assertTrue(method_exists($this->cloudbeds,'delete'));

        $client=$this->mockHttpClient(200);
        $this->cloudbeds->setHttpClient($client);

        $got=$this->cloudbeds->delete('mock_uri','mock_access_token',['mock_param'=>'mock_value']);
        $this->assertIsArray($got);
        $this->assertArrayHasKey('success',$got);
    }
    public function testPut()
    {
        $this->assertTrue(method_exists($this->cloudbeds,'put'));

        $client=$this->mockHttpClient(200);
        $this->cloudbeds->setHttpClient($client);

        $got=$this->cloudbeds->put('mock_uri','mock_access_token',['mock_param'=>'mock_value']);
        $this->assertIsArray($got);
        $this->assertArrayHasKey('success',$got);
    }

    public function testCloudbedsAPIException()
    {
        $this->expectException(CloudbedsAPIException::class);

        $client=$this->mockHttpClient(400,'{"error":true,"message":"mock_error_message"}');
        $this->cloudbeds->setHttpClient($client);
        $this->cloudbeds->get('mock_uri','mock_access_token',['mock_param'=>'mock_value']);
    }

    /**
     * @param $returnCode integer
     * @param string $returnData
     * @return ClientInterface
     */
    private function mockHttpClient($returnCode=200,$returnData='{"success":true}')
    {
        $response = m::mock('Psr\Http\Message\ResponseInterface');
        $response->shouldReceive('getHeader')
            ->times(1)
            ->andReturn('application/json');
        $response->shouldReceive('getStatusCode')
            ->times(1)
            ->andReturn($returnCode);
        $response->shouldReceive('getBody')
            ->times(1)
            ->andReturn($returnData);

        $client = m::mock('GuzzleHttp\ClientInterface');
        $client->shouldReceive('request')
            ->times(1)
            ->withAnyArgs()
            ->andReturn($response);

        return $client;

    }
}

test/src/CloudbedsServiceProviderTest.php
<?php
/**
 * Created by PhpStorm.
 * User: rakib
 * Date: 09-May-19
 * Time: 3:11 AM
 */

namespace R4kib\Cloudbeds\Test;


use Illuminate\Contracts\Foundation\Application;
use PHPUnit\Framework\TestCase;
use R4kib\Cloudbeds\CloudbedsServiceProvider;
use Mockery as m;

class CloudbedsServiceProviderTest extends TestCase
{
    protected $provider;
    protected function setUp()/* The :void return type declaration that should be here would cause a BC issue */
    {
        $app=m::mock(Application::class);
        $app->shouldReceive('singleton')
        ->times(1)
        ->withAnyArgs()
        ->andReturnNull();
        $app->shouldReceive('publishes')
            ->once()
            ->withAnyArgs()
            ->andReturnNull();
        $app->shouldReceive('make')
            ->with('path.config')
            ->once()
            ->andReturn('/some/path');
       $this->provider= new CloudbedsServiceProvider($app);
    }

    public function testCanBeConstructed()
    {
        $this->assertInstanceOf(CloudbedsServiceProvider::class,$this->provider);
    }

    public function testRegister()
    {
        $this->assertNull($this->provider->register());
    }

//    public function testBoot()
//    {
//        $this->assertNull($this->provider->boot());
//    }
    

}

test/src/Provider/CloudbedsProviderTest.php
<?php
/**
 * Created by PhpStorm.
 * User: rakib
 * Date: 06-May-19
 * Time: 2:58 PM
 */

namespace R4kib\Cloudbeds\Test\Provider\CloudbedsTest;

use League\OAuth2\Client\Token\AccessToken;
use PHPUnit\Framework\TestCase;
use R4kib\Cloudbeds\Provider\CloudbedsProvider;
use Mockery as m;
use R4kib\Cloudbeds\Provider\CLoudbedsResourceOwner;

class CloudbedsProviderTest extends TestCase
{
    protected $provider;

    protected function setUp()
    {
        $this->provider = new CloudbedsProvider([
            'clientId' => 'mock_client_id',
            'clientSecret' => 'mock_secret',
            'redirectUri' => 'mock_redirect_uri',
            'apiBaseUrl'=>'https://hotels.cloudbeds.com/api/v1.1'
        ]);
    }

    public function testAuthorizationUrl()
    {
        $url = $this->provider->getAuthorizationUrl();
        $uri = parse_url($url);
        parse_str($uri['query'], $query);
        $this->assertArrayHasKey('client_id', $query);
        $this->assertArrayHasKey('response_type', $query);
        $this->assertArrayHasKey('redirect_uri', $query);
    }

    public function testGetBaseAccessTokenUrl()
    {
        $params = [];
        $url = $this->provider->getBaseAccessTokenUrl($params);
        $uri = parse_url($url);
        $this->assertEquals('/api/v1.1/access_token', $uri['path']);
    }

    public function testGetAuthorizationUrl()
    {
        $url = $this->provider->getAuthorizationUrl();
        $uri = parse_url($url);
        $this->assertEquals('/api/v1.1/oauth', $uri['path']);
    }

    public function testGetResourceOwnerDetailsUrl()
    {
        $url = $this->provider->getResourceOwnerDetailsUrl($this->getAccessToken());
        $uri = parse_url($url);
        $this->assertEquals('/api/v1.1/userinfo', $uri['path']);
    }

    public function testGetAccessToken()
    {
        $client=$this->mockHttpClient('{"access_token":"mock_access_token","refresh_token":"mock_refresh_token","token_type":"bearer","expires_in":3600}');

        $this->provider->setHttpClient($client);
        $token = $this->provider->getAccessToken('authorization_code', ['code' => 'mock_authorization_code']);
        $this->assertEquals('mock_access_token', $token->getToken());
        $this->assertEquals('mock_refresh_token', $token->getRefreshToken());
        $this->assertLessThanOrEqual(time() + 3600, $token->getExpires());
        $this->assertGreaterThanOrEqual(time(), $token->getExpires());
        $this->assertNull($token->getResourceOwnerId(), 'Expected null.');
    }

    public function testGetResourceOwner()
    {
        $client=$this->mockHttpClient('{"user_id":"mock_user_id","first_name":"mock_first_name","last_name":"mock_last_name","email":"mock_email"}');
        $this->provider->setHttpClient($client);

        $owner=$this->provider->getResourceOwner($this->getAccessToken());
        $this->assertInstanceOf(CLoudbedsResourceOwner::class,$owner);
        $this->assertEquals($owner->getId(),'mock_user_id');

    }

    private function mockHttpClient($returnData,$code=200)
    {
        $response = m::mock('Psr\Http\Message\ResponseInterface');
        $response->shouldReceive('getHeader')
            ->times(1)
            ->andReturn('application/json');
        $response->shouldReceive('getStatusCode')
            ->times(1)
            ->andReturn($code);
        $response->shouldReceive('getBody')
            ->times(1)
            ->andReturn($returnData);
        $client = m::mock('GuzzleHttp\ClientInterface');
        $client->shouldReceive('send')->times(1)->andReturn($response);
        return $client;
    }

    private function getAccessToken()
    {
        return new AccessToken(['access_token'=>'mock_token','expires_in'=>time()+3600]);
    }

}

src/CloudbedsServiceProvider.php
<?php

namespace R4kib\Cloudbeds;
use Illuminate\Support\ServiceProvider;

class CloudbedsServiceProvider extends ServiceProvider
{
//    protected $defer = true;
    /**
     * Register services.
     *
     * @return void
     */
    public function register()
    {
        $this->app->singleton(Cloudbeds::class, function()
        {
            return new Cloudbeds([
                'clientId' => config('cloudbeds.api.clientId'),
                'clientSecret' => config('cloudbeds.api.clientSecret'),
                'redirectUri' =>config('cloudbeds.api.redirectUri'),
                'version' => config('cloudbeds.api.version')]) ;
        });
    }

    /**
     * Bootstrap services.
     *
     * @return void
     */
    public function boot()
    {

        $this->publishes([
            __DIR__.'/config/cloudbeds.php' => config_path('cloudbeds.php'),
        ]);
    }
}


src/Provider/CloudbedsProvider.php
<?php
/**
 * Created by PhpStorm.
 * User: rakib
 * Date: 06-May-19
 * Time: 2:36 PM
 */

namespace R4kib\Cloudbeds\Provider;

use League\OAuth2\Client\Provider\AbstractProvider;
use League\OAuth2\Client\Provider\ResourceOwnerInterface;
use Psr\Http\Message\ResponseInterface;
use League\OAuth2\Client\Provider\Exception\IdentityProviderException;
use League\OAuth2\Client\Token\AccessToken;
use League\OAuth2\Client\Tool\BearerAuthorizationTrait;
use R4kib\Cloudbeds\Exceptions\CloudbedsHttpException;

class CloudbedsProvider extends AbstractProvider
{
    private $baseUrl;
    public function __construct(array $options = [], array $collaborators = [])
    {
        parent::__construct($options, $collaborators);
        $this->baseUrl= $options['apiBaseUrl'];

    }

    /**
     * Returns the base URL for authorizing a client.
     * @return string
     */
    public function getBaseAuthorizationUrl()
    {
        return $this->baseUrl."/oauth";
    }

    /**
     * Returns the base URL for requesting an access token.
     * @param array $params
     * @return string
     */
    public function getBaseAccessTokenUrl(array $params)
    {
        return $this->baseUrl.'/access_token';
    }

    /**
     * Returns the URL for requesting the resource owner's details.
     *
     * @param AccessToken $token
     * @return string
     */
    public function getResourceOwnerDetailsUrl(AccessToken $token)
    {
        return $this->baseUrl.'/userinfo';
    }

    /**
     * Returns the default scopes used by this provider.
     *
     * This should only be the scopes that are required to request the details
     * of the resource owner, rather than all the available scopes.
     *
     * @return array
     */
    protected function getDefaultScopes()
    {
        // TODO: Implement getDefaultScopes() method.
    }

    /**
     * Checks a provider response for errors.
     *
     * @param  ResponseInterface $response
     * @param  array|string $data Parsed response data
     * @return void
     * @throws CloudbedsHttpException
     */
    protected function checkResponse(ResponseInterface $response, $data)
    {
        if (isset($data['error'])) {
            throw new CloudbedsHttpException(
                'Cloudbeds API Error ' .  $response->getStatusCode().' - '. $data['message'],
                $response->getStatusCode()
            );
        }
    }

    /**
     * Generates a resource owner object from a successful resource owner
     * details request.
     *
     * @param  array $response
     * @param  AccessToken $token
     * @return CLoudbedsResourceOwner
     */
    protected function createResourceOwner(array $response, AccessToken $token)
    {
        return new CLoudbedsResourceOwner($response);
    }

    protected function getAuthorizationHeaders($token = null)
    {
        return [ 'Authorization' => 'Bearer ' .$token];
    }

    /**
     * Requests and returns the resource owner of given access token.
     *
     * @param  AccessToken $token
     * @return CLoudbedsResourceOwner
*/
    public function getResourceOwner(AccessToken $token)
    {
        $response = $this->fetchResourceOwnerDetails($token);

        return $this->createResourceOwner($response, $token);
    }

}

src/Provider/CLoudbedsResourceOwner.php
<?php
/**
 * Created by PhpStorm.
 * User: rakib
 * Date: 08-May-19
 * Time: 9:10 PM
 */

namespace R4kib\Cloudbeds\Provider;


use League\OAuth2\Client\Provider\ResourceOwnerInterface;
use League\OAuth2\Client\Tool\ArrayAccessorTrait;

class CLoudbedsResourceOwner implements ResourceOwnerInterface
{
    use ArrayAccessorTrait;
    /**
     * Raw response
     *
     * @var array
     */
    protected $response;

    /**
     * CLoudbedsResourceOwner constructor.
     * @param array $response
     */
    public function __construct(array $response = array())
    {
        $this->response = $response;
    }

    /**
     * Returns the identifier of the authorized resource owner.
     *
     * @return mixed
     */
    public function getId()
    {
        return $this->getValueByKey($this->response, 'user_id');
    }

    /**
     * Get resource owner first name
     *
     * @return string|null
     */
    public function getFirstName()
    {
        return $this->getValueByKey($this->response, 'first_name');
    }
    /**
     * Get resource owner last name
     *
     * @return string|null
     */
    public function getLastName()
    {
        return $this->getValueByKey($this->response, 'last_name');
    }

    /**
     * Get resource owner email
     *
     * @return string|null
     */
    public function getEmail()
    {
        return $this->getValueByKey($this->response, 'email');
    }
    /**
     * Return all of the owner details available as an array.
     *
     * @return array
     */
    public function toArray()
    {
        return $this->response;
    }
}

src/Cloudbeds.php
<?php
/**
 * Created by PhpStorm.
 * User: rakib
 * Date: 06-May-19
 * Time: 9:14 PM
 */

namespace R4kib\Cloudbeds;

use GuzzleHttp\Client;
use GuzzleHttp\Exception\GuzzleException;
use League\OAuth2\Client\Token\AccessToken;
use Psr\Http\Message\ResponseInterface;
use R4kib\Cloudbeds\Exceptions\CloudbedsHttpException;
use R4kib\Cloudbeds\Exceptions\CloudbedsAPIException;
use R4kib\Cloudbeds\Exceptions\CloudbedsOperationNotSuccessfulException;


class Cloudbeds
{

    private $options;
    private $baseUrl;
    private $httpClient;

    /**
     * @param mixed $httpClient
     */
    public function setHttpClient($httpClient): void
    {
        $this->httpClient = $httpClient;
    }


    /**
     * Cloudbeds constructor.
     * @param $options array
     */
    public function __construct($options)
    {
        $apiEndpoint = 'https://hotels.cloudbeds.com/api/';
        $version = 'v1.1'; //default version
        if (isset($options['version'])) {
            $version = $options['version'];
        }
        $this->baseUrl = $apiEndpoint . $version;
        $options['apiBaseUrl'] = $this->baseUrl;
        $this->options = $options;
    }

    /**
     * @return array
     */
    public function getOptions(): array
    {
        return $this->options;
    }


    /**
     * @return Provider\CloudbedsProvider
     */
    public function getOauthHelper()
    {
        return new Provider\CloudbedsProvider($this->options);
    }

    public function getHttpClient()
    {
        if (!$this->httpClient) {
            $this->httpClient = $this->getOauthHelper()->getHttpClient();
        }
        return $this->httpClient;
    }

    /**
     * @param $uri string
     * @param $accessToken AccessToken
     * @param $params array
     * @return array
     */
    public function get($uri, $accessToken, $params=[])
    {
        return $this->request('GET',$uri,$accessToken,$params);
    }

    /**
     * @param $uri string
     * @param $accessToken AccessToken
     * @param $params array
     * @return array
     */
    public function post($uri, $accessToken, $params=[])
    {
        return $this->request('POST',$uri,$accessToken,$params);
    }

    /**
     * @param $uri string
     * @param $accessToken AccessToken
     * @param $params array
     * @return array
     */
    public function put($uri, $accessToken, $params=[])
    {
        return $this->request('PUT',$uri,$accessToken,$params);
    }

    /**
     * @param $uri string
     * @param $accessToken AccessToken
     * @param $params array
     * @return array
     */
    public function delete($uri, $accessToken, $params=[])
    {
        return $this->request('DELETE',$uri,$accessToken,$params);
    }

    /**
     * @param $method
     * @param $uri
     * @param $accessToken AccessToken
     * @param $params
     * @return array
     * @throws CloudbedsHttpException
     * @throws CloudbedsAPIException
     */
    private function request($method, $uri, $accessToken, $params=[])
    {
        try {
            $response = $this->getHttpClient()->request($method, $this->baseUrl . $uri, $this->makeOption($accessToken, $params));
        } catch (GuzzleException $e) {
            throw new CloudbedsHttpException($e->getMessage(),$e->getCode(),$e->getPrevious());
        }
        $data=json_decode((string)$response->getBody(), true);
        if (isset($data['error'])) {
            throw new CloudbedsAPIException(
                'Cloudbeds API Error ' .  $response->getStatusCode().' - '. $data['message'],
                $response->getStatusCode()
            );
        }
//        if (isset($data['success']) and $data['success']==false) {
//            throw new CloudbedsOperationNotSuccessfulException(
//                'Operation Failed - '. $data['message'],
//                $response->getStatusCode()
//            );
//        }

        return $data;
    }

    /**
     * @param $accessToken AccessToken
     * @param $params array
     * @return array
     */
    private function makeOption($accessToken, $params)
    {
        $options = ['headers' => [
            'Authorization' => 'Bearer ' . $accessToken->getToken()]];

        if (count($params)>0){
            $options['query'] = $params;
        }
        return $options;
    }

}